Sprint 15

DarkRavenn opened the pull request.

DarkRavenn committed:
Добавлен выбор emoji и цвет трекера

DarkRavenn committed:
Создана модель данных для CoreData

DarkRavenn committed:
- Создан слой работы с БД
- приложение абстрагировано от СoreData

DarkRavenn committed:
Структурные изменения в проекте

DarkRavenn committed:
Подготовка к использованию кор даты

DarkRavenn committed:
Рефакторинг работы с днями недели

DarkRavenn committed:
Исправлен краш на ios 18

DarkRavenn committed:
Реализована CoreData

DarkRavenn committed:
Рефакторинг:
- Ренейминг
- Перемещение файлов

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/Schedule/ScheduleViewController.swift:
```diff
@@ -7,10 +7,11 @@
     private var schedule: [Weekday]
     private let updateSchedule: ([Weekday]) -> Void
     private let tableOptions: [Weekday] = [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday]
+    private let weekdaysText = ["Понедельник", "Вторник", "Среда", "Четверг", "Пятница", "Суббота", "Воскресенье"]
```
Дни недели в текстовом представлении можно поместить внутрь энама Weekday, хранить и брать их оттуда)

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/Schedule/ScheduleViewController.swift:
```diff
@@ -118,20 +119,20 @@ final class SetScheduleViewController: UIViewController {
     }
     
     @objc func doneButtonTapped() {
```
Не забывай, что @objc func **всегда** должны быть приватными

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/Schedule/ScheduleViewController.swift:
```diff
@@ -118,20 +119,20 @@ final class SetScheduleViewController: UIViewController {
     }
     
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! ScheduleTableCell
```
В этих строках есть force операция - force cast, она как и force unwrap является нежелательной и надо стараться такого кода избегать. Можно переписать без их использования, переместив в мягкий каст в if: if let a = a as? 
Всегда когда есть возможность написать безопасную конструкцию – отдавай предпочтение ей. Так твоя работа будет выигрышнее смотреться)

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/EmojiCollectionCell.swift:
```diff
@@ -0,0 +1,49 @@
+    // Идентификатор ячейки — используется для регистрации и восстановления:
+    static let identifier = "EmojiCell"
+    
+    let textLabel: UILabel = {
```
Можно сделать textLabel как private, и менять его внутри ячейки через публичный метод по примеру setEmoji

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/TrackerCreationViewController.swift:
```diff
@@ -13,27 +13,41 @@ struct tableOption {
+    private var selectedColor: IndexPath? = nil
     private var tableOptions: [tableOption] = []
     
+    private let weekdaysText = ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"]
```
И короткую форму тоже можно в энам)

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/TrackerCreationViewController.swift:
```diff
@@ -209,8 +235,12 @@ final class TrackerTypeSelectionViewController: UIViewController {
     }
     
     @objc func createButtonTapped() {
```
И тут можно private

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/TrackerCreationViewController.swift:
```diff
@@ -257,4 +287,119 @@ extension TrackerTypeSelectionViewController: UITableViewDelegate {
+            
+            cell.prepareForReuse()
+            
+            // задаём цвет
```
Лучше удалять такие избыточные комментарии в коде. Я понимаю, что это нужно при работе над проекте, но правилом хорошего тона является чистить код от лишнего перед сдачей работы. В основную ветку проекта должен попадать только чистый код)
Естественно это не касается важных комментариев

andrew-kuzn-yndx commented on Tracker/Helpers/Convertors/ScheduleConvert.swift:
```diff
@@ -0,0 +1,19 @@
+
+    func toWeekdays(_ string: String) -> [Weekday] {
+        if string.isEmpty { return [] }
+        return string.components(separatedBy: ", ").compactMap { Weekday(rawValue: Int($0) ?? 0) }
```
В том числе по этой причине стоит перенести текстовое представление дней недели внутрь энама Weekday, чтобы не делать вот такие дополнительные конвертеры)

DarkRavenn committed:
Правки по ревью:
- Добавлен private для @objc функций
- Убран force cast
- В Weekday добавлены полные и короткие тексстовые значения дня недели
